src/execution_plan/optimizations/optimize_cartesian_product.c

```diff
@@ -0,0 +1,127 @@
+/*
+* Copyright 2018-2020 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "optimize_cartesian_product.h"
+#include "../ops/op_filter.h"
+#include "../ops/op_cartesian_product.h"
+#include "optimizations_util.h"
+
+// Collects all consecutive filters beneath given op.
+static OpFilter **_locate_filters(OpBase *cp) {
+	OpBase *parent = cp->parent;
+	OpFilter **filters = array_new(OpFilter *, 0);
+
+	while(parent && parent->type == OPType_FILTER) {
+		OpFilter *filter_op = (OpFilter *)parent;
+		filters = array_append(filters, filter_op);
+		parent = parent->parent;
+	}
+
+	return filters;
+}
+
+/**
+ * @brief  Creates a structure of dual branched catesian product, followed by a filter.
+ *         The generated record from the cartesian product completely satisfies the filter entities.
+ * @param  *plan: Execution plan to associate its mapping with the new CP operation.
+ * @param  *lhs: Left side branch.
+ * @param  *rhs: Right side branch.
+ * @param  *filter: Trailing filter operation.
+ * @retval Pointer to the filter operation.
+ */
+static OpBase *_chain_cp_and_filter(const ExecutionPlan *plan, OpBase *lhs, OpBase *rhs,
+									OpBase *filter) {
+	OpBase *cp = NewCartesianProductOp(plan);
+	ExecutionPlan_AddOp(cp, lhs);
+	ExecutionPlan_AddOp(cp, rhs);
+	ExecutionPlan_AddOp(filter, cp);
+	return filter;
+}
+
+static void _optimize_cartesian_product(ExecutionPlan *plan, OpBase *cp) {
+	// Retrieve all filter operations located upstream from the Cartesian Product.
+	OpFilter **filter_ops = _locate_filters(cp);
+	uint filter_count = array_len(filter_ops);
+
+	// For each stream joined by the Cartesian product, collect all entities the stream resolves.
+	int stream_count = cp->childCount;
+	rax *stream_entities[stream_count];
+	OptimizeUtils_BuildStreamFromOp(cp, stream_entities, stream_count);
+
+	for(uint i = 0; i < filter_count; i ++) {
+		// Try to create a new dual-branched cartesian product, followed by the current filter.
+		OpFilter *filter_op = filter_ops[i];
+
+		/* A new cartesian product can be generated if both sides of the filter can be fully and
+		 * separately resolved by exactly two child streams. */
+		FT_FilterNode *f = filter_op->filterTree;
```

Comment:
What about: `a.v + b.v == c.v` in a 4 input stream (`a`, `b`, `c` and `d`) CP?
I think the general decision should be: number of streams required to resolve filter < number of input streams.